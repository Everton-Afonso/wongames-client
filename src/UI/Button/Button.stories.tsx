import { Meta, StoryObj } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      control: false
    }
  }
} as Meta

export const Default: StoryObj<ButtonProps> = {}

Default.args = {
  children: 'Buy now'
}

export const WithIcon: StoryObj<ButtonProps> = {}

WithIcon.args = {
  children: 'Buy now',
  icon: <AddShoppingCart />
}
