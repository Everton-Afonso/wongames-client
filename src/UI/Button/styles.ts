'use client'

import styled, { css } from 'styled-components'
import { ButtonProps } from '.'

type WrapperButtonProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'fullWidth'>

const wrapperModifiersButton = {
  small: () => css`
    height: 30px;
    font-size: var(--xsmall);
    padding: 6px 30px;
  `,

  medium: () => css`
    height: 40px;
    font-size: var(--small);
    padding: 10px 32px;
  `,

  large: () => css`
    height: 50px;
    font-size: var(--medium);
    padding: 14px 52px;
  `,

  fullWidth: () => css`
    width: 100%;
  `,

  withIcon: () => css`
    svg {
      width: 18px;

      & + span {
        margin-left: 6px;
      }
    }
  `
}

export const WrapperButton = styled.button<WrapperButtonProps>`
  ${({ size, fullWidth, hasIcon }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border-radius: var(--radius);
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
    color: var(--white);
    border: none;
    padding: 4px 8px;
    cursor: pointer;
    transition: all 0.3s ease;
    text-decoration: none;

    &:hover {
      background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
    }

    ${!!size && wrapperModifiersButton[size]}
    ${!!fullWidth && wrapperModifiersButton.fullWidth()}
    ${!!hasIcon && wrapperModifiersButton.withIcon()}
  `}
`
