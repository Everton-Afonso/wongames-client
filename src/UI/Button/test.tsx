import { render, screen } from '@testing-library/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

describe('<Button />', () => {
  it('should render the button size by default', () => {
    const { container } = render(<Button>Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      padding: '4px 8px'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    render(<Button size="small">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '30px',
      padding: '6px 30px'
    })
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyleRule(
      'font-size',
      'var(--xsmall)'
    )
  })

  it('should render the medium size', () => {
    render(<Button size="medium">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '40px',
      padding: '10px 32px'
    })
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyleRule(
      'font-size',
      'var(--small)'
    )
  })

  it('should render the large size', () => {
    render(<Button size="large">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '50px',
      padding: '14px 52px'
    })
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyleRule(
      'font-size',
      'var(--medium)'
    )
  })

  it('should render a fullWidth version', () => {
    render(<Button fullWidth>Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    render(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    )

    expect(screen.getByText(/buy now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
})
