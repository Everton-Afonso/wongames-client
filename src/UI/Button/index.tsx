import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: React.ReactNode
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  size,
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) => {
  return (
    <S.WrapperButton
      $size={size}
      $fullWidth={fullWidth}
      $hasIcon={!!icon}
      {...props}
    >
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.WrapperButton>
  )
}

export default Button
