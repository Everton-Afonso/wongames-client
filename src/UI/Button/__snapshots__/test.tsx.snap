// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button size by default 1`] = `
.c0 {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: var(--radius);
  background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
  color: var(--white);
  border: none;
  padding: 4px 8px;
  cursor: pointer;
  transition: all 0.3s ease;
  text-decoration: none;
}

.c0:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
